lib/rucio/core/transfer.py

```diff
@@ -702,6 +822,223 @@ def __rewrite_dest_url(dest_url, dest_sign_url, dest_scheme):
+                    'bring_online': bring_online_local,
+                    'copy_pin_lifetime': dict_attributes.get('lifetime', 172800),
+                    'external_host': external_host,
+                    'selection_strategy': 'auto',
+                    'rule_id': rws.rule_id,
+                    'file_metadata': file_metadata,
+                    'dest_scheme_priority': dest_scheme_priority}
+        if len(list_hops) > 1:
+            transfer['multihop'] = True
+            transfer['initial_request_id'] = rws.request_id
+        if strict_copy:
+            transfer['strict_copy'] = strict_copy
+        if use_ipv4:
+            transfer['use_ipv4'] = True
+        if archive_timeout and ctx.is_tape_rse(dest_rse_id):
+            try:
+                transfer['archive_timeout'] = int(archive_timeout)
+                logger(logging.DEBUG, 'Added archive timeout to transfer.')
+            except ValueError:
+                logger(logging.WARNING, 'Could not set archive_timeout for %s. Must be integer.', dest_url)
+                pass
+        if hop is list_hops[-1]:
+            transfer['sources'] = [(source_rse_name, source_url, source_rse_id, source.source_ranking, list_hops[-1]['cumulated_distance'])]
+            transfer['account'] = rws.account
+            if rws.previous_attempt_id:
+                file_metadata['previous_attempt_id'] = rws.previous_attempt_id
+
+        transfer_path.append(transfer)
+    return transfer_path
+
+
+def __pick_best_transfer(candidate_paths):
+    """
+    Given a list of possible transfer paths from different sources towards the same destination,
+    pick the best transfer path.
+    """
+
+    if not candidate_paths:
+        return
+
+    # Discard multihop transfers which contain a tape source as an intermediate hop
+    filtered_candidate_paths = []
+    for path in candidate_paths:
+        if any(transfer['file_metadata']['src_type'] == 'TAPE' for transfer in path[1:]):
+            continue
+        filtered_candidate_paths.append(path)
+    candidate_paths = filtered_candidate_paths
+
+    # Discard multihop transfers which contain other candidate as part of itself For example:
+    # if A->B->C and B->C are both candidates, discard A->B->C because it includes B->C. Doing B->C is enough.
+    source_rses = {path[0]['file_metadata']['src_rse_id'] for path in candidate_paths}
+    filtered_candidate_paths = []
+    for path in candidate_paths:
+        if any(transfer['file_metadata']['src_rse_id'] in source_rses for transfer in path[1:]):
+            continue
+        filtered_candidate_paths.append(path)
+    candidate_paths = filtered_candidate_paths
+
+    def __transfer_order_key(transfer_path):
+        # higher source_ranking first,
+        # on equal source_ranking, prefer DISK over TAPE
+        # on equal type, prefer lower distance_ranking
+        # on equal distance, prefer single hop
+        return (
+            - transfer_path[-1]['sources'][0][3],
```

Comment:
So if there are 2 different transfer paths:
1 - A -> dest where A hosts the file but has lower ranking than B
2 - C -> B -> dest where B doesn't host the file but has higher ranking than A
In that case the multihop will be chosen ?

Reply:
yes, but source_ranking is only decreased at failed transfers. Meaning that A was tried in the past, but failed. So this current transfer is probably result of re-scheduling the failed one.